[DirectX] Use an allow-list of DXIL compatible module metadata

inbelic opened the pull request:
This pr introduces an allow-list for module metadata, this encompasses the llvm metadata nodes: `llvm.ident` and `llvm.module.flags`, as well as, the generated `dx.` options.

Resolves: #164473.

inbelic committed:
[DirectX] Use an allow-list on module metadata

inbelic committed:
style-guide

bob80905 reviewed (approved):
LGTM, just a nit question

Icohedron commented:
There seems to be a failing dxil-dis tests.
I think this one is of interest? `tools/dxil-dis/md-manystrings.ll`
```
2025-10-27T19:01:30.6994330Z ********************
2025-10-27T19:01:30.6994520Z FAIL: LLVM :: tools/dxil-dis/md-manystrings.ll (54511 of 60005)
2025-10-27T19:01:30.6994810Z ******************** TEST 'LLVM :: tools/dxil-dis/md-manystrings.ll' FAILED ********************
2025-10-27T19:01:30.6995040Z Exit Code: 1
2025-10-27T19:01:30.6995100Z 
2025-10-27T19:01:30.6995150Z Command Output (stdout):
2025-10-27T19:01:30.6995320Z --
2025-10-27T19:01:30.6995410Z # RUN: at line 1
2025-10-27T19:01:30.6996850Z /Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/llvm-project/build/bin/llc --filetype=obj /Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/llvm-project/llvm/test/tools/dxil-dis/md-manystrings.ll -o - | /Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/llvm-project/build/bin/dxil-dis | /Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/llvm-project/build/bin/FileCheck /Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/llvm-project/llvm/test/tools/dxil-dis/md-manystrings.ll
2025-10-27T19:01:30.6998850Z # executed command: /Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/llvm-project/build/bin/llc --filetype=obj /Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/llvm-project/llvm/test/tools/dxil-dis/md-manystrings.ll -o -
2025-10-27T19:01:30.6999740Z # executed command: /Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/llvm-project/build/bin/dxil-dis
2025-10-27T19:01:30.7000680Z # executed command: /Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/llvm-project/build/bin/FileCheck /Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/llvm-project/llvm/test/tools/dxil-dis/md-manystrings.ll
2025-10-27T19:01:30.7001340Z # .---command stderr------------
2025-10-27T19:01:30.7001830Z # | /Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/llvm-project/llvm/test/tools/dxil-dis/md-manystrings.ll:42:10: error: CHECK: expected string not found in input
2025-10-27T19:01:30.7002460Z # | ; CHECK: !llvm.too_many_strings = !{!0, !1, !2, !3, !4, !5, !6, !7, !8, !9, !10, !11, !12, !13, !14, !15, !16, !17, !18, !19, !20, !21, !22, !23, !24, !25, !26, !27, !28, !29, !30, !31}
2025-10-27T19:01:30.7002730Z # |          ^
2025-10-27T19:01:30.7002850Z # | <stdin>:1:1: note: scanning from here
2025-10-27T19:01:30.7003120Z # | target datalayout = "e-m:e-p:32:32-i1:8-i8:8-i16:32-i32:32-i64:64-f16:32-f32:32-f64:64-n8:16:32:64"
2025-10-27T19:01:30.7003370Z # | ^
2025-10-27T19:01:30.7003460Z # | 
2025-10-27T19:01:30.7003560Z # | Input file: <stdin>
2025-10-27T19:01:30.7003910Z # | Check file: /Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/llvm-project/llvm/test/tools/dxil-dis/md-manystrings.ll
2025-10-27T19:01:30.7004290Z # | 
2025-10-27T19:01:30.7004420Z # | -dump-input=help explains the following input dump.
2025-10-27T19:01:30.7004600Z # | 
2025-10-27T19:01:30.7004680Z # | Input was:
2025-10-27T19:01:30.7004780Z # | <<<<<<
2025-10-27T19:01:30.7005000Z # |           1: target datalayout = "e-m:e-p:32:32-i1:8-i8:8-i16:32-i32:32-i64:64-f16:32-f32:32-f64:64-n8:16:32:64" 
2025-10-27T19:01:30.7005380Z # | check:42     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
2025-10-27T19:01:30.7005640Z # |           2: target triple = "dxil-ms-dx" 
2025-10-27T19:01:30.7005810Z # | check:42     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2025-10-27T19:01:30.7005960Z # |           3:  
2025-10-27T19:01:30.7006060Z # | check:42     ~
2025-10-27T19:01:30.7006230Z # |           4: !dx.shaderModel = !{!0} 
2025-10-27T19:01:30.7006390Z # | check:42     ~~~~~~~~~~~~~~~~~~~~~~~~
2025-10-27T19:01:30.7006540Z # |           5: !dx.version = !{!1} 
2025-10-27T19:01:30.7006680Z # | check:42     ~~~~~~~~~~~~~~~~~~~~
2025-10-27T19:01:30.7006830Z # |           6: !dx.entryPoints = !{!2} 
2025-10-27T19:01:30.7006980Z # | check:42     ~~~~~~~~~~~~~~~~~~~~~~~~
2025-10-27T19:01:30.7007110Z # |           .
2025-10-27T19:01:30.7007210Z # |           .
2025-10-27T19:01:30.7007300Z # |           .
2025-10-27T19:01:30.7007400Z # | >>>>>>
2025-10-27T19:01:30.7007500Z # `-----------------------------
2025-10-27T19:01:30.7007650Z # error: command failed with exit status: 1
2025-10-27T19:01:30.7007760Z 
2025-10-27T19:01:30.7007800Z --
2025-10-27T19:01:30.7007860Z 
2025-10-27T19:01:30.7007900Z ********************
```

inbelic committed:
update md-manystrings to use DXIL compatible metadata

this test checks that we can embed lots of string as metadata into DXIL
bitcode and retrieve it again

renaming the metadata to llvm.ident retains the testing with a DXIL
compatible metadata node

inbelic committed:
update di-subprogram to only use DXIL compatible metadata

the intent of this test is to ensure that the metadata written with the
DXIL bitcode writer is correctly read back from being embedded in the
DXIL

removing the unrecognized metadata retains the test purpose

inbelic committed:
remove di-subrotine

this test is redundant with respect to di-subprogram, so we can remove
it without reducing test coverage

inbelic committed:
Merge branch 'main' into inbelic/dxil-allowlist

inbelic commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -364,6 +364,16 @@ static void cleanModuleFlags(Module &M) {
+static GlobalMDList CompatibleNamedModuleMDs = {
+    "llvm.ident",     "llvm.module.flags", "dx.resources",   "dx.valver",
+    "dx.shaderModel", "dx.version",        "dx.entryPoints",
+};
```
I agree, unfortunately clang-format does not :(
